### DISCUSSION TITLE
How to install moon scripts

### DISCUSSION THREAD

**Discussion by sorcodiv** [shmuz/far2m#55] [Q&A] [answered]
Как правильно устанавливать moon-скрипты в far2m?
Хочу установить [hexed.moon](https://github.com/shmuz/luafar2m/blob/main/Macros/scripts/Third_party/hexed.moon) ([forum.farmanager.com](https://forum.farmanager.com/viewtopic.php?p=172852#p172852)), копирую скрипт в ~/.config/far2m/Macros/scripts/hexed.moon, но вызвать во вьювере по CtrlF4 не получается - ничего не происходит. Lua-скрипты работают без проблем. Аналогичный скрипт для Far3 (под windows) - работает (располагается по пути Profile\Macros\scripts\hexed.moon).
Пробовал запустить [Dialog.Maximize.moon](https://github.com/shmuz/luafar2m/blob/main/Macros/scripts/Third_party/Dialog.Maximize.moon) - так же нет реакции в диалогах на F2 / CtrlAltRight / CtrlAltLeft.
FAR2M, version 2.6.0-2024-03-06-4b028f2-beta Linux x86_64.

**Comment by shmuz** [comment, marked as answer]
В установке самих скриптов нет разницы между Far3 и far2m.
Разница в том, что в Far3 moonscript идёт вместе с Фаром, а в линуксе его надо установить.
Установка описана [здесь](https://github.com/shmuz/far2m/wiki/MoonScript-installation), но она не всегда удаётся, это зависит от версии luarocks и наличия либо отсутствия в системе других версий Lua.

**Comment by shmuz** [comment]
Проверить, что moonscript правильно установлен, можно командой `lm:about`

![image](https://github.com/shmuz/far2m/assets/344558/6360a4d3-c44a-4824-b32d-c4480a83cca4)

**Comment by sorcodiv** [comment]
Спасибо!
На wiki опечатка, пропущен символ "=":
`sudo luarocks install moonscript --lua-version 5.1`
должно быть:
`sudo luarocks install moonscript --lua-version=5.1`

А каким образом плагин LuaMacro ищет установленный moonscript?
Для эксперимента я скопировал нужные библиотеки в локальный каталог без установки luarocks. Для запуска из консоли можно указать путь с локальным расположением библиотек:
`exec '/usr/bin/luajit' -e 'package.path="./5.1/?.lua;./5.1/?/init.lua;"..package.path; package.cpath="./lib/?.so;"..package.cpath' -e 'local k,l,_=pcall(require,"luarocks.loader") _=k and l.add_context("moonscript","0.5.0-1")' './moon' "$@"`
в консоли на простых примерах moon-скриптов у меня отрабатывает корректно. Но Far при этом не видит, что moonscript установлен. Можно ли плагину LuaMacro указать путь, по которому надо искать интерпретатор moon?

**Comment by shmuz** [comment]
> должно быть:
`sudo luarocks install moonscript --lua-version=5.1`

Там так и было до последних редактирований вики, но я уже запутался - в интернете есть оба варианта.
Кстати на Ubuntu 16.04 luarocks не понимает эту опцию, но без неё прекрасно устанавливает moonscript.

> А каким образом плагин LuaMacro ищет установленный moonscript?

Самым традиционным образом - `require "moonscript"`.
Если в luajit, запущенном в терминале, `require "moonscript"` работает, то и в far2m будет работать.

К сожалению, у меня нет времени экспериментировать с различными способами установки moonscript - но когда-то я успешно установил его на нескольких своих виртуальных машинах, так что это возможно.

> Можно ли плагину LuaMacro указать путь, по которому надо искать интерпретатор moon?

Вроде бы это не предусмотрено.

**Comment by sorcodiv** [comment]
С помощью переменных окружения [LUA_PATH](https://www.lua.org/manual/5.1/manual.html#pdf-package.path) и [LUA_CPATH](https://www.lua.org/manual/5.1/manual.html#pdf-package.cpath) получилось добавить поддержку moonscript в [portable-сборку](https://github.com/sorcodiv/far2lua-build/releases). Проверил на разных дистрибутивах на примере hexed.moon - вроде работает. Спасибо!

Кстати, похоже нашел баг: https://github.com/shmuz/far2m/issues/59

**Reply by shmuz** [reply]
Отлично, рад, что у вас получилось.

Сам я в данное время не пользуюсь сборками portable, т.к. у меня на всех виртуалках (и на одном не виртуальном) линуксе far2m собирается из исходников.

Кстати, любопытно, чем отличаются между собой ваша сборка и [вот эта](https://github.com/Zeroes1/far2m-portable) ?

**Comment by sorcodiv** [comment]
> Кстати, любопытно, чем отличаются между собой ваша сборка и [вот эта](https://github.com/Zeroes1/far2m-portable) ?

Принципиально - ничем :) Скрипты для сборки основаны на одном [источнике](https://github.com/unxed/far2l-deb) (unxed более не делает portable-сборок).

**Comment by shmuz** [comment]
Для информации:
Начиная от коммита c765ac68 MoonScript входит в состав far2m, его больше не надо отдельно устанавливать.
